File: code/src/main/python/store/mongo_driver.py
import sys
import os

sys.path.append(os.path.abspath("."))
sys.dont_write_bytecode = True

__author__ = "COSAL"


import pymongo
import multiprocessing

from utils import cache, logger


LOGGER = logger.get_logger(os.path.basename(__file__.split(".")[0]))
MAX_POOL_SIZE=10000
__MONGO_CLIENTS = {}


def get_default_hostname():
  # return socket.gethostname()
  return 'localhost'


def get_hostname():
  mongo_home = os.environ.get("MONGO_HOME", None)
  if mongo_home:
    file_name = os.path.join(mongo_home, "host_machine.txt")
    if cache.file_exists(file_name):
      return cache.read_file(file_name).strip()
  return get_default_hostname()


def get_client():
  global __MONGO_CLIENTS
  process_id = multiprocessing.current_process().ident
  if process_id not in __MONGO_CLIENTS:
    __MONGO_CLIENTS[process_id] = pymongo.MongoClient(host=get_hostname(), maxPoolSize=MAX_POOL_SIZE)
  __MONGO_CLIENT = __MONGO_CLIENTS[process_id]
  try:
    assert __MONGO_CLIENT.address
    return __MONGO_CLIENT
  except Exception:
    __MONGO_CLIENT.close()
    __MONGO_CLIENT = None
    del __MONGO_CLIENTS[process_id]
    raise RuntimeError("Looks like client is down")


def get_dataset_db(dataset):
  if not dataset:
    raise RuntimeError("Dataset is empty")
  try:
    return get_client().get_database(dataset)
  except Exception as e:
    LOGGER.warn("Dataset '%s' does not exist. Creating this mongo manually and rerun "
                "the application. Don't forget to insert dummy record" % dataset)
    raise RuntimeError(e)


def get_collection(dataset, collection_name):
  return get_dataset_db(dataset).get_collection(collection_name)


def is_collection_exists(collection):
  return collection.estimated_document_count() > 0


def create_unique_index_for_collection(collection, *fields):
  indices = []
  for field in fields:
    indices.append((field, pymongo.ASCENDING))
  collection.create_index(indices, unique=True)


def create_index_for_collection(collection, field, order=pymongo.ASCENDING):
  if does_index_exist(collection, field):
    # LOGGER.info("Index: '%s' exists for collection: '%s'. Not creating ..." % (field, collection.name))
    return
  # LOGGER.info("Creating index: '%s' for collection: '%s' ..." % (field, collection.name))
  collection.create_index([(field, order)])


def does_index_exist(collection, field):
  index_info = collection.index_information()
  if not index_info:
    return False
  for key, info in index_info.items():
    for row in info['key']:
      if row[0] == field:
        return True
  return False


def get_document(collection, key, value):
  if collection is None:
    return None
  return collection.find_one({key: value})


def contains_document(collection, key, value):
  if collection is None:
    return False
  return collection.find_one({key: value}, {"_id": 1})


def upsert_document(collection, query, document):
  collection.update_one(
    query,
    {"$set": document},
    upsert=True
  )


def delete_document(collection, key, value):
  collection.delete_one({key: value})


def delete_document_with_query(collection, query):
  collection.delete_one(query)


def delete_documents(collection, key, value):
  collection.delete_many({key: value})


def delete_documents_with_query(collection, query):
  collection.delete_many(query)


def drop_collection(collection):
  collection.drop()


def _test_index_info():
  print(does_index_exist(get_collection("Misconceptions", "differences"), "d_jaroq"))


if __name__ == "__main__":
  # print(get_dataset_db("codejam"))
  _test_index_info()

File: code/src/main/java/edu/ncsu/store/mongo/MongoDriver.java
package edu.ncsu.store.mongo;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import com.mongodb.MongoClient;
import com.mongodb.client.FindIterable;
import com.mongodb.client.MongoCollection;
import com.mongodb.client.MongoDatabase;
import com.mongodb.client.model.Filters;
import com.mongodb.client.model.IndexOptions;
import com.mongodb.client.model.Indexes;
import com.mongodb.client.result.UpdateResult;
import com.mongodb.util.JSON;
import edu.ncsu.utils.Utils;
import org.bson.Document;
import org.bson.conversions.Bson;

import java.net.InetAddress;
import java.net.UnknownHostException;
import java.util.Map;
import java.util.logging.Level;
import java.util.logging.Logger;

public class MongoDriver {

    private static final Logger LOGGER = Logger.getLogger(MongoDriver.class.getName());

    private static final Logger mongoLogger = Logger.getLogger( "org.mongodb.driver" );

    private static Gson GSON = new GsonBuilder().serializeSpecialFloatingPointValues().create();

    private static MongoClient client = null;

    private final static JsonParser jsonParser = new JsonParser();

    static {
         mongoLogger.setLevel(Level.OFF);
    }

    private static String getDefaultHostName() {
        return "localhost";
//        try {
//            return InetAddress.getLocalHost().getHostName();
//        } catch (UnknownHostException e) {
//            LOGGER.severe(String.format("Failed to fetch host name. Exception: %s", e.getMessage()));
//            throw new RuntimeException(e);
//        }
    }

    public static String getHostName() {
        String mongoHome = System.getenv("MONGO_HOME");
        if ( mongoHome != null && !mongoHome.isEmpty()) {
            String fileName = Utils.pathJoin(mongoHome, "host_machine.txt");
            String hostName = Utils.getFileContent(fileName);
            if (hostName != null && !hostName.isEmpty()) {
                return hostName;
            }
        }
        return getDefaultHostName();
    }

    public static MongoClient getClient() {
        if (client == null) {
            String hostName = MongoDriver.getHostName();
            client = new MongoClient(hostName);
        }
        try {
            client.getAddress();
        } catch (Exception e) {
            LOGGER.severe("Looks like client is down");
            client.close();
            client = null;
            throw new RuntimeException(e);
        }
        return client;
    }

    public static MongoDatabase getDatasetDB(String dataset) {
        if (dataset == null || dataset.isEmpty()) {
            throw new RuntimeException("Dataset is empty");
        }
        try {
            return getClient().getDatabase(dataset);
        } catch (NullPointerException e) {
            LOGGER.severe(String.format("Dataset '%s' does not exist. Creating this mongo manually " +
                    "and rerun the application. Don't forget to insert a dummy record.", dataset));
            throw e;
        }

    }

    public static MongoCollection<Document> getCollection(String dataset, String collectionName) {
        return getDatasetDB(dataset).getCollection(collectionName);
    }

    @SuppressWarnings("rawtypes")
    public static void createIndexForCollection(MongoCollection collection, String... fields) {
        IndexOptions indexOptions = new IndexOptions().unique(true);
        collection.createIndex(Indexes.ascending(fields), indexOptions);
    }

    @SuppressWarnings("rawtypes")
    public static void createNonUniqueIndexForCollection(MongoCollection collection, String field) {
        collection.createIndex(Indexes.ascending(field));
    }

    @SuppressWarnings("rawtypes")
    public static boolean collectionExists(MongoCollection collection) {
        return collection != null && collection.count() > 0;
    }

    public static void dropCollection(String dataset, String collectionName) {
        getDatasetDB(dataset).getCollection(collectionName).drop();
    }

    @SuppressWarnings("rawtypes")
    public static Document getDocument(MongoCollection collection, String key, Object value) {
        return getDocument(collection, Filters.eq(key, value));
    }

    @SuppressWarnings("rawtypes")
    public static boolean containsDocument(MongoCollection collection, String key, Object value) {
        return getDocument(collection, key, value) != null;
    }

    @SuppressWarnings("rawtypes")
    public static void deleteDocument(MongoCollection collection, String key, Object value) {
        deleteDocument(collection, Filters.eq(key, value));
    }

    @SuppressWarnings("rawtypes")
    public static Document getDocument(MongoCollection collection, Bson query) {
        return (Document) collection.find(query).first();
    }

    @SuppressWarnings("rawtypes")
    public static FindIterable getDocuments(MongoCollection collection, Bson query) {
        return collection.find(query);
    }

    @SuppressWarnings("rawtypes")
    public static FindIterable getDocuments(MongoCollection collection, Bson query, Bson projections) {
        return collection.find(query).projection(projections);
    }

    @SuppressWarnings("rawtypes")
    public static long countDocuments(MongoCollection collection, Bson query) {
        return collection.count(query);
    }

    @SuppressWarnings("rawtypes")
    public static boolean containsDocument(MongoCollection collection, Bson query) {
        return getDocument(collection, query) != null;
    }

    @SuppressWarnings("rawtypes")
    public static void deleteDocument(MongoCollection collection, Bson query) {
        collection.deleteMany(query);
    }

    @SuppressWarnings("rawtypes")
    public static void updateDocument(MongoCollection collection, Bson query, Bson update) {
        if (containsDocument(collection, query)) {
            UpdateResult result = collection.updateOne(query, update);
            if (!result.wasAcknowledged())
                throw new RuntimeException("@COSAL: Check why update failed!!");
        }
    }

    public static void dropDatabase(String databaseName) {
        MongoDatabase db = getDatasetDB(databaseName);
        if (db != null)
            db.drop();
    }

    @SuppressWarnings("unchecked")
    public static Document parseAsDocument(JsonObject result) {
        String jsonString = GSON.toJson(result).trim().replaceAll("(-)?Infinity", "NaN");
        return new Document((Map<String, Object>) JSON.parse(jsonString));
    }

    public static JsonObject parseAsJson(Document document) {
        String jsonString = document.toJson();
        return jsonParser.parse(jsonString).getAsJsonObject();
    }

    public static void main(String[] args) {
        for (Document doc: getClient().listDatabases()) {
            System.out.println(doc.get("name"));
        }
//        System.out.println(getClient().listDatabases());
//        LOGGER.info(String.format("HostName of MongoD Server: %s", MongoDriver.getClient().getAddress().getHost()));
    }

}
